[#43] GCM 기능 초기 설계

dannyjoo7 opened the pull request:
### 개요
- GCM 기능 초기 설계

### 작업내용
1. GCM 기능 패키지 분할
2. GCM 기능 초기 설계
3. Chat 관련 UI 로딩 bar 추가

### 주요 변경점
- FCM -> GCM

---------

close #43

dannyjoo7 committed:
[#43] gradle 설정 변경
- fcm 추가

dannyjoo7 committed:
[#43] manifest 설정 변경
- fcm service 추가

dannyjoo7 committed:
[#43] colors.xml 색 추가
- wd2_main_color 추가

dannyjoo7 committed:
[#43] merge conflict 해결

dannyjoo7 committed:
[#43] 패키지 분할로 인한 변경

dannyjoo7 committed:
[#43] fcm 기초 코드 추가

dannyjoo7 committed:
[#43] ChatFragment loading 시 Progressbar 추가

dannyjoo7 committed:
[#43] ChatDetailActivity loading 시 Progressbar 추가

dannyjoo7 committed:
[#43] 변수명 변경
 - loadingState -> isLoading

dannyjoo7 committed:
[#43] Private면 txtMemberNum 숨김 처리

dannyjoo7 committed:
[#43] 오브젝트명 변경 및 추가
 - RetrofitClient -> KAKAORetrofitClient로 변경
 - GCMRetrofitClient 추가

dannyjoo7 committed:
[#43] GCM을 이용한 push 알림 전송 기능 추가
 - GCMRequestInterceptor 추가
 - GCMService 추가
 - ChatRepositoryImpl.addChatMessageItem 내 push 알림 기능 추가

dannyjoo7 committed:
[#43] String 값 변경

sinw212 commented on app/src/main/java/com/wd/woodong2/data/repository/ChatRepositoryImpl.kt:
```diff
@@ -132,6 +140,32 @@ class ChatRepositoryImpl(private val databaseReference: DatabaseReference) : Cha
+        // FCM Notification 객체 생성
+        // to -> 받는 사람 Token
+        val notification = FCMNotification(
+            to = "fO_rDAHQR3CBrr4q0gqg3h:APA91bGyFjStWQF76yCG0a406LM3gF7NVolK76Ht1qLMq_Ht-FAgd4SQetZDfuNlPlFTS9lFpE3geef-mXvjELodyk2Pqxld_j1V6Ip1s1YSgZfNy6YTei7uX9ivEQylHpRMF3rJSjVN",
```
받는 사람 Token은 테스트용으로 넣어둔 값인가요~? string resource로 빼서 사용하는게 좋지 않을까요!

dannyjoo7 replied:
[c896f30](https://github.com/TeamWD2/WooDong2/pull/49/commits/c896f308d6669d6aa958bfe44c6616e5491fd89e)
찾아보니 String 값으로 사용하는 것 추천하더라구요! 감사합니다!

sinw212 commented on app/src/main/java/com/wd/woodong2/presentation/chat/content/ChatFragment.kt:
```diff
@@ -77,6 +78,9 @@ class ChatFragment : Fragment() {
         chatViewModel.chatList.observe(viewLifecycleOwner) { itemList ->
             chatItemListAdapter.submitList(itemList.toMutableList())
         }
+        chatViewModel.isLoading.observe(viewLifecycleOwner) { loadingState ->
+            binding.progressBar.isVisible = loadingState
+        }
```
로딩바 구현 좋네요👍

dannyjoo7 replied:
감사합니다~

sinw212 commented on app/src/main/java/com/wd/woodong2/retrofit/RetrofitClient.kt:
RetrofitClient 코드를 각각 작성하는것에 대해서는 확인이 필요할 것 같아요!

dannyjoo7 replied:
BASE_URL이 달라 부득이하게 따로 각각 작성하게 되었습니다~

sinw212 commented on app/src/main/java/com/wd/woodong2/services/fcm/GCMService.kt:
GCM 역시 Retrofit 통신을 통해 데이터를 가져오는거니까 Data Layer의 Remote Source로 사용하는게 좋지않을까요~?

dannyjoo7 replied:
[416b21b](https://github.com/TeamWD2/WooDong2/pull/49/commits/416b21bdd52ff759562c5c00862365952053ea52)
[31a0a38](https://github.com/TeamWD2/WooDong2/pull/49/commits/31a0a384920b60f4bd5701d4f2ce23324a52672b)
[ae6f58b](https://github.com/TeamWD2/WooDong2/pull/49/commits/ae6f58b6ac7befee110159011e33c6a3fe41d732)
전체적으로 파일명을 수정하고 레이어도 알맞게 옮겨놨습니다~ 감사합니다!

sinw212 commented on app/src/main/java/com/wd/woodong2/services/fcm/MyFirebaseMessagingService.kt:
```diff
@@ -0,0 +1,91 @@
+            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE
+        )
+
+        // 액션을 넘길 수 있다(클래스 이름) manifest에 정의해서 Action을 통해 특정 액티비티를 실행시킬 수 있는 방법
+        val notificationBuilder = NotificationCompat.Builder(this, "chat_message_id")
+            .setSmallIcon(R.drawable.wd2)
+            .setContentTitle(noti?.title)
+            .setContentText(noti?.body)
+            .setSound(RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)) // 알림 사운드 설정
+            .setContentIntent(pendingIntent)
+            .setAutoCancel(true)
+
+        val notificationManager =
+            getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+
+        // Notification Channel 생성 (Android 8.0 이상에서 필요)
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            val channel = NotificationChannel(
+                "chat_message_id",
+                "Chat Message",
+                NotificationManager.IMPORTANCE_DEFAULT
+            )
+            notificationManager.createNotificationChannel(channel)
+        }
+        notificationManager.notify(0, notificationBuilder.build())
+    }
```
함수로 분리해두면 가독성이 좋아질 것 같아요~!

dannyjoo7 replied:
[ae6f4c6](https://github.com/TeamWD2/WooDong2/pull/49/commits/ae6f4c6d91dd1bbcdb5f77ea7b373301533937ef)
감사합니다~ 반영했습니다!

dannyjoo7 committed:
[#43] 피드백 반영
 - Test Token 값 하드코딩 -> R.string.test_client_token 값 변경

dannyjoo7 committed:
[#43] 피드백 반영 : 파일 분할 및 이름 변경
 - RetrofitClient -> KAKAORetrofitClient, GCMRetrofitClient로 분할
 - RequestInterceptor -> KAKAORequestInterceptor 이름 변경

dannyjoo7 committed:
[#43] 피드백 반영 : 레이어 위치 변경 및 이름 변경
 - service.fcm -> data.remotesource 이동
 - GCMService -> GCMRemoteSource 변경

dannyjoo7 committed:
[#43] 피드백 반영 : 레이어 위치 변경 및 이름 변경
 - FCMItem -> GCMRequest 파일명 변경
 - services.fcm -> data.model 위치 변경

dannyjoo7 committed:
[#43] 피드백 반영 : noti 메소드 분할
 - createNotification 생성

sinw212 reviewed (approved):
피드백 반영 확인했습니다~
